keps/sig-api-machinery/storage-migration-auto-trigger.md

```diff
@@ -225,6 +208,137 @@ manually, e.g., deploying the [initializer][] manually.
 
+## Extended API proposal for HA cluster
+
+The proposed storage state and triggering controller are not guaranteed to work
+during the rolling upgrade of HA masters. The issue is detailed in the previous
+[KEP][]. Essentially, the migrator should abort ongoing storage version
+migrations when the HA masters are rolling upgrading, and restart the migration
+when the rolling upgrade is done. We propose the following changes to the
+storage state API and the triggering controller to detect if there is ongoing
+master rolling upgrade.
+
+[KEP]:35-storage-version-hash.md#ha-masters
+
+### Extending the storage state API
+
+```golang
+// Remaining the same.
+type StorageState struct {
+  metav1.TypeMeta
+  metav1.ObjectMeta
+  Spec StorageStateSpec
+  Status StorageStateStatus
+}
+
+// Remaining the same.
+type StorageStateSpec {
+  Resource GroupResource
+}
+
+// Extended.
+type StorageStateStatus {
+  // Extended.
+  // A ist of current storage version hashes. In case of HA setup, the
+  // storage version hashes exposed by all master instances are reported here.
+  // In case of non-HA setup, this list should only contain one element.
+  CurrentStorageVersionHashes []StorageVersionHash
+  // Remaining the same.
+  // +optional
+  PersistedStorageVersionHashes []string
+}
+
+// The storage version hash observed by the reporter at lastHeartbeatTime.
+type StorageVersionHash struct {
+  // The ID of the reporting triggering contoller. This should be the same ID
+  // that's used to contend for a lease, see https://github.com/kubernetes/kubernetes/blob/428a8e04d40ef01c28c66fcfd54f306a1aff9a28/staging/src/k8s.io/api/coordination/v1/types.go#L43.
+  // Required.
+  ReporterID string
+  // The hash value of the current storage version, as shown in the discovery
+  // document served by the API server local to the reporting controller.
+  CurrentStorageVersionHash string
+  // LastHeartbeatTime is the last time the reporting controller checks the
+  // storage version hash of this resource in the discovery document and updates
+  // this field.
+  // +optional
+  LastHeartbeatTime metav1.Time
+}
+``` 
+
+### Extending the work flow of the triggering controller
+
+Normally, HA controllers select a leader, and only the leader writes to the API
+to update status. In this case, we need every triggering controller to report
+its local storage version hash. Specifically, every 10 minutes, each instance of
+the triggering controller 
+* fetches the discovery doc from the local apiserver.
+* adds the storage version hash to PersistedStorageVersionHashes if the local
+  storage version hash is not present there yet.
+* updates (or creates) the element in status.storageVersionhashes with the
+  matching reporterID, recording the latest currentStorageVersionHash and
+  lastHeartbeatTime.
+* removes storageVersionHashes whose lastHeartbeatTime are more than 2 polling
+  periods (20 mins) old. This garbage collection is necessary because the
```

Comment:
Is 2 polling periods enough? If the controller dies immediately before updating the status (or didn't manage to update it due to some other issues), then that kind of introduces a race between next try and GC.
As an example, we're marking node as NotReady 40s after now heartbeat, whereas heartbeat period is 10s (so 4x heartbeat time).
I would suggest increasing 2x to something larger (at least 3x, if not 4x) to avoid strange situations.

Reply:
Yeah, increasing the GC grace period is reasonable. Increased to 3x.